File: Examples/ore-rest-api/src/main/java/com/ore/api/service/TemplateService.java
package com.ore.api.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.ore.api.dto.OREInputs;
import com.ore.api.dto.Parameters;
import com.ore.api.dto.Setup;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.thymeleaf.TemplateEngine;
import org.thymeleaf.context.Context;
import org.springframework.core.io.ResourceLoader;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Service for rendering ORE XML templates using Thymeleaf.
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class TemplateService {

    private final TemplateEngine templateEngine;
    private final ResourceLoader resourceLoader;
    private final ObjectMapper objectMapper;

    @Value("${ore.templates.dir:classpath:templates}")
    private String templatesDir;

    @Value("${ore.shared-input-dir:../Input}")
    private String sharedInputDir;

    /**
     * Render ORE input files for a job.
     */
    public Path renderJobInputs(
            String templateName,
            OREInputs inputs,
            Path jobInputDir,
            Path jobOutputDir,
            String jobType
    ) {
        try {
            // Build template context
            Context context = buildContext(templateName, inputs, jobInputDir, jobOutputDir);

            // Write user-provided files
            writeUserFiles(inputs, jobInputDir);

            // Render ore.xml
            String templatePath = templateName + "/ore";
            String oreXml = templateEngine.process(templatePath, context);

            Path oreXmlPath = jobInputDir.resolve("ore.xml");
            Files.writeString(oreXmlPath, oreXml);

            log.info("Rendered ore.xml for template '{}'", templateName);
            return oreXmlPath;

        } catch (Exception e) {
            throw new RuntimeException("Failed to render template: " + templateName, e);
        }
    }

    private Context buildContext(
            String templateName,
            OREInputs inputs,
            Path jobInputDir,
            Path jobOutputDir
    ) {
        Map<String, Object> defaults = loadDefaults(templateName);

        Context context = new Context();

        // Build setup section
        @SuppressWarnings("unchecked")
        Map<String, Object> setup = (Map<String, Object>) defaults.getOrDefault("setup", new HashMap<>());
        setup = new HashMap<>(setup);
        setup.put("asOfDate", inputs.asOfDate());
        if (inputs.baseCurrency() != null) {
            setup.put("baseCurrency", inputs.baseCurrency());
        }

        // Compute relative path to shared input directory
        String sharedPath = computeRelativePath(jobInputDir, Path.of(sharedInputDir));
        setup.put("sharedInputDir", sharedPath);

        // Override shared file paths
        String[] sharedFiles = {
            "marketDataFile", "fixingDataFile", "curveConfigFile",
            "conventionsFile", "marketConfigFile", "pricingEnginesFile"
        };

        for (String key : sharedFiles) {
            Object value = setup.get(key);
            if (value instanceof String str && str.startsWith("../../Input/")) {
                String filename = str.substring("../../Input/".length());
                setup.put(key, sharedPath + "/" + filename);
            }
        }

        // Apply user parameters if provided
        if (inputs.parameters() != null && inputs.parameters().setup() != null) {
            applyUserOverrides(setup, inputs.parameters().setup());
        }

        context.setVariable("setup", setup);

        // Markets section
        @SuppressWarnings("unchecked")
        Map<String, String> markets = (Map<String, String>) defaults.getOrDefault("markets", new HashMap<>());
        if (inputs.parameters() != null && inputs.parameters().markets() != null) {
            markets.putAll(inputs.parameters().markets());
        }
        context.setVariable("markets", markets);

        // Analytics section
        @SuppressWarnings("unchecked")
        Map<String, Map<String, Object>> analytics =
            (Map<String, Map<String, Object>>) defaults.getOrDefault("analytics", new HashMap<>());

        if (inputs.parameters() != null && inputs.parameters().analytics() != null) {
            for (var entry : inputs.parameters().analytics().entrySet()) {
                analytics.computeIfAbsent(entry.getKey(), k -> new HashMap<>())
                    .putAll((Map<String, Object>) entry.getValue());
            }
        }
        context.setVariable("analytics", analytics);

        // Output path
        String outputPath = computeRelativePath(jobInputDir, jobOutputDir);
        context.setVariable("outputPath", outputPath);

        return context;
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> loadDefaults(String templateName) {
        try {
            String defaultsPath = "classpath:templates/" + templateName + "/defaults.yml";
            var resource = resourceLoader.getResource(defaultsPath);

            if (!resource.exists()) {
                log.debug("No defaults.yml found for template '{}'", templateName);
                return new HashMap<>();
            }

            // Parse YAML (simple approach - in production, use SnakeYAML)
            // For now, return empty map and let templates handle defaults
            return new HashMap<>();

        } catch (Exception e) {
            log.warn("Failed to load defaults for template '{}': {}", templateName, e.getMessage());
            return new HashMap<>();
        }
    }

    private void writeUserFiles(OREInputs inputs, Path jobInputDir) throws IOException {
        Files.createDirectories(jobInputDir);

        if (inputs.portfolioFile() != null) {
            Files.writeString(jobInputDir.resolve("portfolio.xml"), inputs.portfolioFile());
        }
        if (inputs.marketData() != null) {
            Files.writeString(jobInputDir.resolve("market.txt"), inputs.marketData());
        }
        if (inputs.fixingData() != null) {
            Files.writeString(jobInputDir.resolve("fixings.txt"), inputs.fixingData());
        }
        if (inputs.curveConfig() != null) {
            Files.writeString(jobInputDir.resolve("curveconfig.xml"), inputs.curveConfig());
        }
        if (inputs.conventions() != null) {
            Files.writeString(jobInputDir.resolve("conventions.xml"), inputs.conventions());
        }
        if (inputs.marketConfig() != null) {
            Files.writeString(jobInputDir.resolve("todaysmarket.xml"), inputs.marketConfig());
        }
        if (inputs.pricingEngines() != null) {
            Files.writeString(jobInputDir.resolve("pricingengine.xml"), inputs.pricingEngines());
        }
        if (inputs.simulationConfig() != null) {
            Files.writeString(jobInputDir.resolve("simulation.xml"), inputs.simulationConfig());
        }
        if (inputs.nettingSets() != null) {
            Files.writeString(jobInputDir.resolve("netting.xml"), inputs.nettingSets());
        }
    }

    private String computeRelativePath(Path from, Path to) {
        try {
            Path fromAbs = from.toAbsolutePath();
            Path toAbs = to.toAbsolutePath();

            Path relative = fromAbs.relativize(toAbs);
            String path = relative.toString().replace("\\", "/");

            // Count how many levels up we need to go
            int levels = 0;
            Path current = fromAbs;
            while (current != null && !current.startsWith(toAbs)) {
                current = current.getParent();
                levels++;
            }

            if (levels > 0) {
                StringBuilder sb = new StringBuilder();
                for (int i = 0; i < levels; i++) {
                    sb.append("..");
                    if (i < levels - 1) sb.append("/");
                }
                return sb.toString();
            }

            return path.isEmpty() ? "." : path;
        } catch (IllegalArgumentException e) {
            // Can't compute relative path (different drives on Windows)
            return to.toAbsolutePath().toString();
        }
    }

    private void applyUserOverrides(Map<String, Object> target, Setup userSetup) {
        if (userSetup.inputPath() != null) target.put("inputPath", userSetup.inputPath());
        if (userSetup.outputPath() != null) target.put("outputPath", userSetup.outputPath());
        if (userSetup.logFile() != null) target.put("logFile", userSetup.logFile());
        if (userSetup.logMask() != null) target.put("logMask", userSetup.logMask());
    }

    public List<String> listTemplates() {
        try {
            var resource = resourceLoader.getResource("classpath:templates/");
            if (resource.exists()) {
                var file = resource.getFile();
                if (file.isDirectory()) {
                    return List.of(file.list((dir, name) -> new java.io.File(dir, name).isDirectory()));
                }
            }
        } catch (IOException e) {
            log.warn("Failed to list templates: {}", e.getMessage());
        }

        // Return default templates
        return List.of("pricing-basic", "xva-standard", "stress-ir-up", "stress-ir-down");
    }
}

File: Examples/ORE-RESTAPI/app/services/template_service.py
import logging
import os
from pathlib import Path
from typing import Any

import yaml
from jinja2 import Environment, FileSystemLoader

from app.config import settings
from app.schemas.job import OREInputs

logger = logging.getLogger(__name__)


class TemplateNotFoundError(Exception):
    pass


class TemplateService:
    """Renders ORE XML input files from Jinja2 templates and user inputs."""

    def __init__(self):
        self.templates_dir = Path(settings.TEMPLATES_DIR)
        self.shared_input_dir = Path(settings.SHARED_INPUT_DIR)
        self.jinja_env = Environment(
            loader=FileSystemLoader(str(self.templates_dir)),
            autoescape=False,
            keep_trailing_newline=True,
        )

    def render_job_inputs(
        self,
        template_name: str,
        inputs: OREInputs,
        job_input_dir: Path,
        job_output_dir: Path,
    ) -> Path:
        """Render all ORE input files for a job.

        Returns the path to the generated ore.xml.
        """
        # Validate template exists
        template_dir = self.templates_dir / template_name
        if not template_dir.exists():
            raise TemplateNotFoundError(f"Template not found: {template_name}")

        # Load defaults
        defaults = self._load_defaults(template_name)

        # Merge user inputs into defaults
        context = self._build_context(defaults, inputs, job_input_dir, job_output_dir)

        # Write user-provided supporting files
        self._write_user_files(inputs, job_input_dir)

        # Render ore.xml
        template = self.jinja_env.get_template(f"{template_name}/ore.xml.j2")
        ore_xml = template.render(**context)

        ore_xml_path = job_input_dir / "ore.xml"
        ore_xml_path.write_text(ore_xml, encoding="utf-8")
        logger.info("Rendered ore.xml for template '%s'", template_name)

        return ore_xml_path

    def _load_defaults(self, template_name: str) -> dict[str, Any]:
        """Load default values from templates/{name}/defaults.yml."""
        defaults_path = self.templates_dir / template_name / "defaults.yml"
        if defaults_path.exists():
            with open(defaults_path, "r", encoding="utf-8") as f:
                return yaml.safe_load(f) or {}
        return {}

    def _build_context(
        self,
        defaults: dict[str, Any],
        inputs: OREInputs,
        job_input_dir: Path,
        job_output_dir: Path,
    ) -> dict[str, Any]:
        """Build the Jinja2 template context from defaults + user inputs."""
        context: dict[str, Any] = {}

        # Determine file references first
        # Shared files use relative path from job input dir to Examples/Input/
        # Use os.path.relpath to handle cross-tree relative paths
        shared_prefix = os.path.relpath(
            str(self.shared_input_dir.resolve()),
            str(job_input_dir.resolve())
        ).replace("\\", "/")

        # Setup section
        setup = defaults.get("setup", {}).copy()
        setup["asofDate"] = inputs.asOfDate
        if inputs.baseCurrency:
            setup["baseCurrency"] = inputs.baseCurrency

        # Override shared file paths with computed relative paths
        # Only override if the value starts with "../../Input/" (from template defaults)
        shared_file_keys = [
            "marketDataFile", "fixingDataFile", "curveConfigFile",
            "conventionsFile", "marketConfigFile", "pricingEnginesFile"
        ]
        for key in shared_file_keys:
            if key in setup and setup[key].startswith("../../Input/"):
                filename = setup[key].replace("../../Input/", "")
                setup[key] = f"{shared_prefix}/{filename}"

        # Override setup params from user
        if inputs.parameters and "setup" in inputs.parameters:
            setup.update(inputs.parameters["setup"])
        context["setup"] = setup

        # Markets section
        markets = defaults.get("markets", {}).copy()
        if inputs.parameters and "markets" in inputs.parameters:
            markets.update(inputs.parameters["markets"])
        context["markets"] = markets

        # Analytics section
        analytics = defaults.get("analytics", {}).copy()
        if inputs.parameters and "analytics" in inputs.parameters:
            for key, val in inputs.parameters["analytics"].items():
                if key in analytics:
                    analytics[key].update(val)
                else:
                    analytics[key] = val
        context["analytics"] = analytics

        context["shared_input_dir"] = shared_prefix

        # User-provided files are in the job input directory (no prefix needed)
        context["job_input_dir"] = "."
        # Output is a sibling of input: jobs/{id}/output relative to jobs/{id}/input
        context["output_path"] = str(
            job_output_dir.resolve().relative_to(job_input_dir.resolve().parent)
        ).replace("\\", "/")

        return context

    def _write_user_files(self, inputs: OREInputs, job_input_dir: Path) -> None:
        """Write user-provided file contents to the job input directory."""
        file_map = {
            "portfolioFile": inputs.portfolioFile,
            "marketData": inputs.marketData,
            "fixingData": inputs.fixingData,
            "curveConfig": inputs.curveConfig,
            "conventions": inputs.conventions,
            "marketConfig": inputs.marketConfig,
            "pricingEngines": inputs.pricingEngines,
            "simulationConfig": inputs.simulationConfig,
            "nettingSets": inputs.nettingSets,
        }
        filenames = {
            "portfolioFile": "portfolio.xml",
            "marketData": "market.txt",
            "fixingData": "fixings.txt",
            "curveConfig": "curveconfig.xml",
            "conventions": "conventions.xml",
            "marketConfig": "todaysmarket.xml",
            "pricingEngines": "pricingengine.xml",
            "simulationConfig": "simulation.xml",
            "nettingSets": "netting.xml",
        }

        for key, content in file_map.items():
            if content:
                filepath = job_input_dir / filenames[key]
                filepath.write_text(content, encoding="utf-8")
                logger.info("Wrote user file: %s", filenames[key])

    def list_templates(self) -> list[str]:
        """List available template names."""
        if not self.templates_dir.exists():
            return []
        return [
            d.name
            for d in self.templates_dir.iterdir()
            if d.is_dir() and (d / "ore.xml.j2").exists()
        ]


template_service = TemplateService()
